conformance/native/build.gradle.kts

```diff
@@ -0,0 +1,15 @@
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+
+plugins {
+    kotlin("multiplatform")
+    application
```

Comment:
❓ Are you able to actually run the native conformance binary? I thought the `application` plugin is only for bundling JVM-based projects and you would need to do something like what's at https://kotlinlang.org/docs/reference/building-mpp-with-gradle.html#building-final-native-binaries for building the native version of the conformance binary.

Reply:
🤦‍♂️
You are right...  I'll take a look at the link, thanks!

Reply:
After a bit of investigation, I can tell you that the main entry point of an executable binary can't be a suspend function. And that's the case with the main function of `Main.kt`. I don't know why this is a suspend function. I'll make the change.

Reply:
That makes sense. I think you should be able to just change:

```kt
suspend fun main(args: Array<String>) {
```
to:

```kt
fun main(args: Array<String>) = runBlocking<Unit> {
```

Reply:
I tried that but I got an error on that line:
`val res = doTestIo().also { debug { "Result: $it" } } ?: return`

Because that function can't return 🤔 I'll have to dig deeper

Reply:
Also, `runBlocking` is not available on JS targets...

Reply:
Yeah I just noticed that too. That's a bit unfortunate. I've been playing around with how we can get rid of the `suspend` on `main`. The below patch seems to work in both JVM and JS on `master`. Can you try it out with Native and see if it solves your problem?

```patch
diff --git a/conformance/lib/src/commonMain/kotlin/pbandk/conformance/Main.kt b/conformance/lib/src/commonMain/kotlin/pbandk/conformance/Main.kt
index e47ebfd..e65cf5b 100644
--- a/conformance/lib/src/commonMain/kotlin/pbandk/conformance/Main.kt
+++ b/conformance/lib/src/commonMain/kotlin/pbandk/conformance/Main.kt
@@ -1,15 +1,18 @@
 package pbandk.conformance
 
 import pbandk.ByteArr
+import pbandk.conformance.Platform.runBlocking
 import pbandk.conformance.pb.*
 
 var logDebug = false
-inline fun debug(fn: () -> String) { if (logDebug) Platform.stderrPrintln(fn()) }
+inline fun debug(fn: () -> String) {
+    if (logDebug) Platform.stderrPrintln(fn())
+}
 
-suspend fun main(args: Array<String>) {
+fun main(args: Array<String>) = runBlocking {
     debug { "Starting conformance test" }
     while (true) {
-        val res = doTestIo().also { debug { "Result: $it" } } ?: return
+        val res = doTestIo().also { debug { "Result: $it" } } ?: return@runBlocking
         ConformanceResponse(res).protoMarshal().also { bytes ->
             Platform.stdoutWriteIntLE(bytes.size)
             Platform.stdoutWriteFull(bytes)
diff --git a/conformance/lib/src/commonMain/kotlin/pbandk/conformance/Platform.kt b/conformance/lib/src/commonMain/kotlin/pbandk/conformance/Platform.kt
index 20892cd..4386616 100644
--- a/conformance/lib/src/commonMain/kotlin/pbandk/conformance/Platform.kt
+++ b/conformance/lib/src/commonMain/kotlin/pbandk/conformance/Platform.kt
@@ -1,5 +1,7 @@
 package pbandk.conformance
 
+import kotlinx.coroutines.CoroutineScope
+
 expect object Platform {
     fun stderrPrintln(str: String)
 
@@ -10,4 +12,6 @@ expect object Platform {
     fun stdoutWriteFull(arr: ByteArray)
 
     inline fun <T> doTry(fn: () -> T, errFn: (Any) -> T): T
+
+    fun runBlocking(block: suspend CoroutineScope.() -> Unit)
 }
\ No newline at end of file
diff --git a/conformance/lib/src/jsMain/kotlin/pbandk/conformance/Platform.kt b/conformance/lib/src/jsMain/kotlin/pbandk/conformance/Platform.kt
index 631c364..5a5b4b8 100644
--- a/conformance/lib/src/jsMain/kotlin/pbandk/conformance/Platform.kt
+++ b/conformance/lib/src/jsMain/kotlin/pbandk/conformance/Platform.kt
@@ -1,5 +1,6 @@
 package pbandk.conformance
 
+import kotlinx.coroutines.*
 import org.khronos.webgl.ArrayBufferView
 import org.khronos.webgl.Uint8Array
 import pbandk.asByteArray
@@ -104,4 +105,8 @@ actual object Platform {
         } catch (e: dynamic) {
             errFn(e as Any)
         }
+
+    actual fun runBlocking(block: suspend CoroutineScope.() -> Unit) {
+        GlobalScope.launch(block = block)
+    }
 }
\ No newline at end of file
diff --git a/conformance/lib/src/jvmMain/kotlin/pbandk/conformance/Platform.kt b/conformance/lib/src/jvmMain/kotlin/pbandk/conformance/Platform.kt
index 686c8fe..d41d583 100644
--- a/conformance/lib/src/jvmMain/kotlin/pbandk/conformance/Platform.kt
+++ b/conformance/lib/src/jvmMain/kotlin/pbandk/conformance/Platform.kt
@@ -1,5 +1,6 @@
 package pbandk.conformance
 
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.withContext
 import java.nio.ByteBuffer
@@ -27,5 +28,13 @@ actual object Platform {
 
     actual fun stdoutWriteFull(arr: ByteArray) = System.out.write(arr)
 
-    actual inline fun <T> doTry(fn: () -> T, errFn: (Any) -> T) = try { fn() } catch (e: Exception) { errFn(e) }
+    actual inline fun <T> doTry(fn: () -> T, errFn: (Any) -> T) = try {
+        fn()
+    } catch (e: Exception) {
+        errFn(e)
+    }
+
+    actual fun runBlocking(block: suspend CoroutineScope.() -> Unit) {
+        kotlinx.coroutines.runBlocking(block = block)
+    }
 }
```

Reply:
Haha 😆I was working on something like that!! I'll check the patch and have something working today or tomorrow.

But at first, it's working on native without any problem

Reply:
🎉 Wouhou!! I have the conformance tests running!!! 😃

But I have some unexpected failures...

Reply:
@garyp Do you have any tips for investigating failed conformance tests? Thanks 😄

Reply:
Do you have a list of which ones are failing?

BTW, looks like the `test-conformance.sh` script has a bug. It ignores failures for all platforms except for the last one in the script. So even though CircleCI is reporting the build for this PR as green, it actually did fail in the `conformance_native` job: https://app.circleci.com/pipelines/github/streem/pbandk/105/workflows/fb38e944-cdc0-4fdb-9b39-5352c312542a/jobs/135/parallel-runs/0/steps/0-103.

I'll get the script fixed. But in the meantime, can you take a look at why that job is failing? Once it's running correctly I can look at the logs and help you figure out the failing conformance tests.

Reply:
Those are the tests that are failing:
```
Required.Proto2.ProtobufInput.RepeatedScalarSelectsLast.STRING.ProtobufOutput
Required.Proto2.ProtobufInput.ValidDataRepeated.STRING.ProtobufOutput  Required.Proto2.ProtobufInput.ValidDataScalar.STRING[6].ProtobufOutput
Required.Proto3.ProtobufInput.RepeatedScalarSelectsLast.STRING.ProtobufOutput
Required.Proto3.ProtobufInput.ValidDataRepeated.STRING.ProtobufOutput
Required.Proto3.ProtobufInput.ValidDataScalar.STRING[6].ProtobufOutput
```